### ISSUE TITLE
Document AVX2 implications of numcodecs dependency

### ISSUE THREAD

**Issue by brianzhang01** [tskit-dev/tsinfer#177]
tl;dr: I suggest adding some documentation similar to that found in numcodecs [here](https://numcodecs.readthedocs.io/en/stable/#installation) describing AVX2 optimisations. I'm happy to add this and suggest it goes on [this page](https://tsinfer.readthedocs.io/en/latest/installation.html).

What follows is a slightly self-indulgent narrative of how I debugged this issue...

For my use case, I was setting up tsinfer to run on a cluster, specifically Oxford's Wellcome Trust Research Computing cluster ("ResComp"). I used `python -m pip install --user tsinfer` from a login node to install tsinfer, and things worked fine on the login node.

When I then submitted jobs to the cluster using `qsub`, I got intermittent failures at `import tsinfer` with a message like
```
/var/spool/rescomp/compc045/job_scripts/17662211: line 1: 20440 Illegal instruction     python3 "$@"
```

This was not a very helpful error message. I first Googled this line and the [top hit](https://stackoverflow.com/questions/49094597/illegal-instruction-core-dumped-after-running-import-tensorflow) turned out to be related, about AVX instructions and pre-compiled TensorFlow binaries.

By printing out `cat /proc/cpuinfo | grep flags` across different runs, I found that when the runs succeeded, the machines had these additional flags: `fma movbe abm tsc_adjust bmi1 avx2 bmi2 invpcid cqm cqm_llc cqm_occup_llc`. I began to suspect AVX2 instructions as the culprit.

From here I was stuck for some time, but finally realized I could [debug Python using GDB](https://wiki.python.org/moin/DebuggingWithGdb). This took some time to set up, but once it did, I got the more helpful message:
```
Program received signal SIGILL, Illegal instruction.
__Pyx_modinit_type_init_code () at numcodecs/vlen.c:20608
20608	numcodecs/vlen.c: No such file or directory.
```

Eventually I ended up on the [numcodecs documentation](https://numcodecs.readthedocs.io/en/stable/), which thankfully had the description about AVX2 on its front page. In summary, the login nodes I was using had AVX2 support, so when I compiled with pip, numcodecs specialized to use AVX2 instructions. The fix was to uninstall numcodecs, run
```
$ export DISABLE_NUMCODECS_AVX2=
$ pip install -v --no-cache-dir numcodecs
```
as described on their documentation, and then I installed tsinfer again using pip.

**Comment by jeromekelleher** [comment]
Ouch, sorry about this @brianzhang01. Yes, please do update the docs with a warning about this. I guess this will be alleviated a bit when we ship conda binaries for tsinfer.

**Comment by hyanwong** [comment]
@brianzhang01 - I'm going through tsinfer docs now, so if you want to send me some text, I can incorporate it. Otherwise just make your own PR.

**Comment by brianzhang01** [comment]
That would be great, as I'm also about to head off on holiday -- thanks.

In the section [Installation](https://tsinfer.readthedocs.io/en/latest/installation.html), we can use this text.

Firstly, "Please the msprime installation documentation for details on the various to address this problem." --> "Please **see** the msprime installation documentation for details on the various **ways** to address this problem."

Immediately following in a new paragraph: "Note that one of the dependencies of `tsinfer`, `numcodecs`, specialises to use AVX2 instructions if those are available on your hardware. This has led to issues for [some users](https://github.com/tskit-dev/tsinfer/issues/177) when installing on hardware with AVX2, then trying to use `tsinfer` from a cluster compute node without AVX2. To resolve this, `numcodecs` has a `DISABLE_NUMCODECS_AVX2` variable which can be turned on before calling `pip install`, see [these instructions](https://numcodecs.readthedocs.io/en/stable/#installation
) for details."

If you want to edit the text down further, perfectly fine by me. As long as we link to this issue users should be able to follow the trail of links back to the numcodecs documentation. Also, feel free to merge without me taking a look, thanks!
